tests/utils/test_log_handlers.py

```diff
@@ -317,43 +317,52 @@ def test__read_for_celery_executor_fallbacks_to_worker(self, create_task_instanc
-        ti.try_number = 2
-        with conf_vars({("core", "executor"): executor_name}):
-            reload(executor_loader)
-            fth = FileTaskHandler("")
-
-            fth._read_from_logs_server = mock.Mock()
-            fth._read_from_logs_server.return_value = ["this message"], ["this\nlog\ncontent"]
-            actual = fth._read(ti=ti, try_number=2)
-            fth._read_from_logs_server.assert_called_once()
-            assert actual == ("*** this message\nthis\nlog\ncontent", {"end_of_log": False, "log_pos": 16})
-
-            # Previous try_number should return served logs when remote logs aren't implemented
-            fth._read_from_logs_server = mock.Mock()
-            fth._read_from_logs_server.return_value = ["served logs try_number=1"], ["this\nlog\ncontent"]
-            actual = fth._read(ti=ti, try_number=1)
-            fth._read_from_logs_server.assert_called_once()
-            assert actual == (
-                "*** served logs try_number=1\nthis\nlog\ncontent",
-                {"end_of_log": True, "log_pos": 16},
+        # Reading logs from worker should occur when the task is either running, deferred, or up for retry.
+        for state in (TaskInstanceState.RUNNING, TaskInstanceState.DEFERRED, TaskInstanceState.UP_FOR_RETRY):
+            ti = create_task_instance(
+                dag_id=f"dag_for_testing_celery_executor_log_read_{state}",
+                task_id="task_for_testing_celery_executor_log_read",
+                run_type=DagRunType.SCHEDULED,
+                execution_date=DEFAULT_DATE,
             )
-
-            # When remote_logs is implemented, previous try_number is from remote logs without reaching worker server
-            fth._read_from_logs_server.reset_mock()
-            fth._read_remote_logs = mock.Mock()
-            fth._read_remote_logs.return_value = ["remote logs"], ["remote\nlog\ncontent"]
-            actual = fth._read(ti=ti, try_number=1)
-            fth._read_remote_logs.assert_called_once()
-            fth._read_from_logs_server.assert_not_called()
-            assert actual == ("*** remote logs\nremote\nlog\ncontent", {"end_of_log": True, "log_pos": 18})
+            ti.try_number = 2
+            ti.state = state
+            with conf_vars({("core", "executor"): executor_name}):
+                reload(executor_loader)
+                fth = FileTaskHandler("")
+                fth._read_from_logs_server = mock.Mock()
+                fth._read_from_logs_server.return_value = ["this message"], ["this\nlog\ncontent"]
+                actual = fth._read(ti=ti, try_number=2)
+                fth._read_from_logs_server.assert_called_once()
+                # If we are in the up for retry or deferred state, log has ended and a new task try has started.
+                # When the state is running, the log has not ended.
+                expected_end_of_log = state in (TaskInstanceState.DEFERRED, TaskInstanceState.UP_FOR_RETRY)
```

Comment:
Noticed while updating the tests that when the state is either of these, the `ti.try_number` is automatically incremented, making it slightly difficult to test the behavior correctly.

Is `TaskInstanceState.DEFERRED` a state where the current attempt can also be deferred/paused? If so, I'm not sure if this test is accurate, but writing more tests for that case is challenging as per above.

Reply:
I.e. the tests the way they were previously or even after these changes don't fail with the following change to logic

```sh
diff --git a/airflow/utils/log/file_task_handler.py b/airflow/utils/log/file_task_handler.py
index 6f1c02e931..bd5cbe9127 100644
--- a/airflow/utils/log/file_task_handler.py
+++ b/airflow/utils/log/file_task_handler.py
@@ -407,7 +407,7 @@ class FileTaskHandler(logging.Handler):
         )
         log_pos = len(logs)
         messages = "".join([f"*** {x}\n" for x in messages_list])
-        end_of_log = ti.try_number != try_number or not is_in_running_or_deferred
+        end_of_log = ti.try_number != try_number or ti.state != TaskInstanceState.RUNNING
         if metadata and "log_pos" in metadata:
             previous_chars = metadata["log_pos"]
             logs = logs[previous_chars:]  # Cut off previously passed log test as new tail

```
This caused some headache while debugging.